_posts/gRPC-Go performance improvements.md

```diff
@@ -0,0 +1,148 @@
+
+However, this begs at least two questions: 1. Won't a client be able to send as much data as it wants to the server by creating new streams when one runs out? 2. Why even have connection-level flow control if the stream-level flow control is enough.
+
+The answer to the first question is short and simple: no. A server has an option to limit the number of streams that it intends to server concurrently. Therefore, although at first it may seem like a problem, it really is not.
+
+The second question is very interesting: why even bother with connection-level flow control?
+
+It is true that stream-level flow control is sufficient to throttle a sender from sending too much data. But not having connection-level flow control (or using an unlimited connection-level window) makes it so that when things get slower on a stream, opening a new one will appear to make things faster. This is a slippery slope and will only take one so far. However, having a connection-level flow control window set to the Bandwidth Delay Product (BDP) of the network puts an upper-bound on how much performance can realistically be squeezed out of the network.
+
+
+### Piggyback window updates
+
+[Code link](https://github.com/grpc/grpc-go/pull/1273)
+
+Sending a window update itself has a cost associated to it; a flush operation is necessary, which results in a syscall. Syscalls are blocking and slow. Therefore, when sending out a stream-level window update, it makes sense to also check if a connection-level window update can be sent using the same flush syscall.
+
+
+### BDP estimation and dynamic flow control window
+
+[Code link](https://github.com/grpc/grpc-go/pull/1310)
+
+This feature  is the latest and in some ways the most awaited optimization feature that has helped us close the final gap between gRPC and HTTP/1.1 performance on high latency networks.
+
+Bandwidth Delay Product (BDP) is the bandwidth of a network connection times its round-trip latency.  This effectively tells us how many bytes can be "on the wire" at a given moment, if full utilization is achieved.
+
+The algorithm to compute BDP and adapt accordingly was first proposed by @ejona and later implemented by both gRPC-C core and gRPC-Java (note that it isn't enabled in Java yet). The idea is simple and powerful: every time a receiver gets a data frame it sends out a BDP ping (a ping with unique data only used by BDP estimator). After this, the receiver starts counting the number of bytes it receives (including the ones that triggered the BDP ping) until it receives the ack for that ping. This total sum of all bytes received in about 1.5 RTT (Round-Trip Time) is an approximation of the effective BDP * 1.5. If this is close to our current window size (say, more than 2/3rd of it) we must increase the window size. We put our window sizes (both streaming and connection) to be twice the BDP we sampled.
```

Comment:
Let's introduce the definition of 'sample' here since it's referred to subsequently.
The sample size (total sum of all bytes received),